astropy/units/physical.py

```diff
@@ -1,133 +1,473 @@
+    physical types may be used for dimensional analysis.
+
+    >>> length = u.pc.physical_type
+    >>> area = (u.cm ** 2).physical_type
+    >>> length * area
+    PhysicalType('volume')
+    >>> area / length
+    PhysicalType('length')
+    >>> length ** 3
+    PhysicalType('volume')
+
+    Unknown physical types are labelled as ``"unknown"``.
+
+    >>> (u.s ** 13).physical_type
+    PhysicalType('unknown')
+
+    Dimensional analysis may be performed for unknown physical types too.
+
+    >>> length_to_19th_power = (u.m ** 19).physical_type
+    >>> length_to_20th_power = (u.m ** 20).physical_type
+    >>> length_to_20th_power / length_to_19th_power
+    PhysicalType('length')
+    """
+
+    def __init__(self, unit, physical_types):
+        self._unit = _replace_temperatures_with_kelvin(unit)
+        self._physical_type_id = self._unit._get_physical_type_id()
+        self._physical_type = _standardize_physical_type_names(physical_types)
+        self._physical_type_list = sorted(self._physical_type)
+
+    def __iter__(self):
+        yield from self._physical_type_list
+
+    def __eq__(self, other):
+        """
+        Return `True` if ``other`` represents a physical type that is
+        consistent with the physical type of the `PhysicalType` instance.
+        """
+        if isinstance(other, PhysicalType):
+            return self._physical_type_id == other._physical_type_id
+        elif isinstance(other, str):
+            other = _standardize_physical_type_names(other)
+            return other.issubset(self._physical_type)
+        else:
+            return NotImplemented
+
+    def __ne__(self, other):
+        equality = self.__eq__(other)
+        return not equality if isinstance(equality, bool) else NotImplemented
+
+    @property
+    def _name_string_as_ordered_set(self):
+        return "{" + str(self._physical_type_list)[1:-1] + "}"
+
+    def __repr__(self):
+        if len(self._physical_type) == 1:
+            names = "'" + self._physical_type_list[0] + "'"
+        else:
+            names = self._name_string_as_ordered_set
+        return f"PhysicalType({names})"
+
+    def __str__(self):
+        if len(self._physical_type) == 1:
+            return self._physical_type_list[0]
+        else:
+            return self._name_string_as_ordered_set
+
+    @staticmethod
+    def _dimensionally_compatible_unit(obj):
+        """
+        If a unit is passed in, return that unit.  If a physical type is
+        passed in, return a unit that corresponds to that physical type.
+        If a real number is passed in, return a dimensionless unit.
+        Otherwise, return `NotImplemented`.
+        """
+        if isinstance(obj, core.UnitBase):
+            return _replace_temperatures_with_kelvin(obj)
+        elif isinstance(obj, PhysicalType):
+            return obj._unit
+        elif isinstance(obj, numbers.Real):
+            return core.dimensionless_unscaled
+        else:
+            return NotImplemented
```

Comment:
Maybe add a test that something like `length.__mul__([])` returns `NotImplemented`.